### ISSUE TITLE
Fatal error: exception Retyping.RetypeError(0)

### ISSUE THREAD

**Issue by shlevy** [coq/coq#13093]
#### Description of the problem

Compiling:

```coq
CoInductive conat :=
| co0 : conat
| coS : conat -> conat.

Definition corec
           {B}
           (x : forall {A} (f : A) (g' : B -> A), B -> A) : B -> conat :=
  cofix corec' (b : B) : conat :=
    let
      f : conat := co0
    in let
      g' : B -> conat := fun b => coS (corec' b)
    in
    x f g' b.
```

gives:

```
$ coqc Test.v 
File "./Test.v", line 8, characters 2-144:
Error:
<in exception printer>:<original exception:Anomaly
                                           "Uncaught exception Retyping.RetypeError(0)."
Please report at http://coq.inria.fr/bugs/.

Fatal error: exception Retyping.RetypeError(0)
```

#### Coq Version

```
$ coqtop -v # ignore the timestamp, compiled in a deterministic environment
The Coq Proof Assistant, version 8.12.0 (January 1980)
compiled on Jan 1 1980 0:00:00 with OCaml 4.10.0
```

Running on Linux.

**Comment by ejgallego** [comment]
Funny one:
```
Error:
<in exception printer>:
Raised at file "pretyping/retyping.ml", line 51, characters 22-44
Called from file "pretyping/retyping.ml", line 225, characters 19-58
Called from file "pretyping/retyping.ml", line 215, characters 26-46
Called from file "pretyping/detyping.ml", line 886, characters 18-64
Called from file "camlinternalLazy.ml", line 27, characters 17-27
Re-raised at file "camlinternalLazy.ml", line 34, characters 4-11
Called from file "interp/constrextern.ml", line 833, characters 38-48
Called from file "interp/constrextern.ml", line 941, characters 34-57
Called from file "interp/constrextern.ml", line 1005, characters 14-64
Called from file "printing/proof_diffs.ml", line 253, characters 37-96
Called from file "vernac/himsg.ml", line 94, characters 34-57
Called from file "vernac/himsg.ml" (inlined), line 431, characters 35-62
Called from file "vernac/himsg.ml", line 500, characters 31-57
Called from file "vernac/himsg.ml", line 1380, characters 11-16
Called from file "lib/cErrors.ml", line 99, characters 10-13
Called from file "lib/cErrors.ml", line 112, characters 4-49
<original exception:
Anomaly "Uncaught exception Retyping.RetypeError(0)."
Please report at http://coq.inria.fr/bugs/.
Raised at file "kernel/type_errors.ml", line 139, characters 2-68
Called from file "kernel/inductive.ml", line 1358, characters 8-99
Called from file "pretyping/pretyping.ml", line 766, characters 17-61
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "pretyping/pretyping.ml", line 770, characters 15-39
Called from file "pretyping/pretyping.ml" (inlined), line 1320, characters 2-81
Called from file "pretyping/pretyping.ml", line 1341, characters 21-85
Called from file "pretyping/pretyping.ml", line 1386, characters 20-71
Called from file "interp/constrintern.ml", line 2437, characters 17-65
Called from file "vernac/comDefinition.ml", line 104, characters 31-100
Called from file "vernac/comDefinition.ml", line 119, characters 4-71
Called from file "vernac/vernacentries.ml", line 606, characters 6-123
Called from file "vernac/vernacinterp.ml", line 53, characters 13-18
Called from file "vernac/vernacinterp.ml", line 213, characters 24-64
Called from file "lib/flags.ml", line 17, characters 14-17
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "vernac/vernacinterp.ml", line 262, characters 18-43
Called from file "vernac/vernacinterp.ml", line 260, characters 6-279
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "stm/stm.ml", line 2391, characters 16-43
Called from file "stm/stm.ml", line 957, characters 6-10
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "stm/stm.ml", line 2532, characters 4-105
Called from file "stm/stm.ml", line 2713, characters 4-60
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "toplevel/vernac.ml", line 68, characters 31-52
Re-raised at file "clib/exninfo.ml", line 82, characters 4-38
Called from file "toplevel/coqloop.ml", line 428, characters 17-62
Called from file "toplevel/coqloop.ml", line 465, characters 4-39
Called from file "toplevel/coqloop.ml", line 471, characters 6-34
```

**Comment by Alizter** [comment]
This seems to be fixed in 8.15 alpha. I am getting a warning:
```coq

Recursive definition of corec' is ill-formed.
In environment
B : ?X1
x : forall A : Type, A -> (B -> A) -> B -> A
corec' : B -> conat
b : B
Sub-expression "let f := co0 in let g' := fun b : B => coS (corec' b) in x conat f g' b" not in guarded
form (should be a constructor, an abstraction, a match, a cofix or a recursive
call).
Recursive definition is:
"fun b : B => let f := co0 in let g' := fun b0 : B => coS (corec' b0) in x conat f g' b".
```

**Comment by ejgallego** [comment]
Thanks for triaging @Alizter , let's close then.

**Comment by herbelin** [comment]
Probably fixed by PR #13108 (Oct 2020), in which case the cause is similar to #13154.
